File: Leetcode-Java/SortList.java
import java.util.LinkedList;

/**
 * Created by yhf on 15/6/15.
 */

/*
Sort a linked list in O(n log n) time using constant space complexity.
*/

public class SortList {
    public static ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode fast = head, slow = head;
        while (fast.next != null && fast.next.next != null) {   // 注意这里不能写 fast != null && fast.next != null
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode head2 = slow.next;
        slow.next = null;
        return mergeLists(sortList(head), sortList(head2));
    }

    private static ListNode mergeLists(ListNode head1, ListNode head2) {
        ListNode head = new ListNode(0);
        ListNode cur = head;
        while (head1 != null && head2 != null) {
            if (head1.val <= head2.val) {
                cur.next = head1;
                head1 = head1.next;
            } else {
                cur.next = head2;
                head2 = head2.next;
            }
            cur = cur.next;
        }
        if (head1 != null) {
            cur.next = head1;
        }
        if (head2 != null) {
            cur.next = head2;
        }
        return head.next;
    }

    public static void main(String[] args) {
        ListNode n1 = new ListNode(5);
        ListNode n2 = new ListNode(4);
        ListNode n3 = new ListNode(3);
        ListNode n4 = new ListNode(2);
        ListNode n5 = new ListNode(1);
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;
        n4.next = n5;
        ListNode head = sortList(n1);
    }
}

File: Leetcode-Python/SortList.py
"""
Sort a linked list in O(n log n) time using constant space complexity.
"""
from ListNode import ListNode


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def sortList(self, head):
        if not head or not head.next:
            return head
        fast = slow = head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        new_head = slow.next
        slow.next = None
        head1 = self.sortList(head)
        head2 = self.sortList(new_head)
        return self.merge_lists(head1, head2)

    def merge_lists(self, head1, head2):
        if not head1 or not head2:
            return head1 or head2
        dummy_head = cur = ListNode(0)
        while head1 and head2:
            if head1.val <= head2.val:
                cur.next = head1
                head1 = head1.next
            else:
                cur.next = head2
                head2 = head2.next
            cur = cur.next
        if head1 or head2:
            cur.next = head1 or head2
        return dummy_head.next


if __name__ == '__main__':
    print Solution().sortList(ListNode(5, 4, 3, 2, 1))
